doc: revamp macOS build doc

jarolrod opened the pull request:
This PR makes the macOS build-docs more informative and adds in the following information:
- Proper descriptions and delineation of required/optional dependencies
- walk-through of optional dependencies
- configuration walk-through
- various other tidbits of information

This is a part of the efforts done in https://github.com/bitcoin/bitcoin/pull/20601 and https://github.com/bitcoin/bitcoin/pull/20610 to update the docs and introduce some consistency between them.

This update does not add instructions for arm-based M1 Macbooks as I do not have one to test with. It would be nice to have someone follow up with an update containing instructions for arm-based Macs.

**Before/Master:** [render](https://github.com/bitcoin/bitcoin/blob/master/doc/build-osx.md)
**After/PR:** [render](https://github.com/bitcoin/bitcoin/blob/c180c911b88b2bd2baf2c9c2b24e276787ffb69b/doc/build-osx.md)

jarolrod committed:
doc: revamp macOS build doc

This pr makes the macOS build docs more informative and adds in the following information:
- Proper descriptions and delineation of required/optional dependencies
- walk-through of optional dependencies
- configuration walk-through
- various other tid-bits of information

hebasto commented:
Why dropped "Usually, macOS installation already has a suitable SQLite installation." ?

jarolrod commented:
> Why dropped "Usually, macOS installation already has a suitable SQLite installation." ?

Since it was introduced in 10.14, which is our minimum supported macOS, I guess you're right in that it cant stay

hebasto commented on doc/build-osx.md:
````diff
@@ -1,116 +1,316 @@
+Skip if you don't intend to use legacy wallets.
+
+``` bash
+brew install berkeley-db4
````
Package name `berkeley-db4` seems [old](https://github.com/Homebrew/formulae.brew.sh/blob/master/_data/formula/berkeley-db%404.json). Maybe use (here and other places) `berkeley-db@4`?

jarolrod replied:
You're right: https://formulae.brew.sh/formula/berkeley-db@4

Using the `Formerly Known as` name might not be so safe

hebasto commented on doc/build-osx.md:
```diff
@@ -1,116 +1,316 @@
+[automake](https://formulae.brew.sh/formula/automake)      | Build      | Generate makefile
+[libtool](https://formulae.brew.sh/formula/libtool)        | Build      | Shared library support
+[pkg-config](https://formulae.brew.sh/formula/pkg-config)  | Build      | Configure compiler and linker flags
+[git](https://formulae.brew.sh/formula/git)                | Clone      | Version control system
```
Technically, `git` is not a build dependency :)

jarolrod commented:
updated from 6d08ec2 -> 4780ea7, addressed @hebasto suggestions

changes:
- use new 'safe' name for berekeley-db
- git is not exactly a build dependency, so it has been removed
- Various little text-fix ups found upon another read through

fanquake commented on doc/build-osx.md:
```diff
@@ -1,116 +1,319 @@
+
+Library                                                    | Purpose    | Description
+-----------------------------------------------------------|------------|----------------------
+[Xcode Tools](https://developer.apple.com/xcode/features/) | Build      | macOS build tools
```
What are the Xcode Tools? Having this here is going to make users think they need to install Xcode. Which unless it's changed recently, is not required.

jarolrod replied:
removed in c180c911b88b2bd2baf2c9c2b24e276787ffb69b

fanquake commented on doc/build-osx.md:
````diff
@@ -1,116 +1,319 @@
 ```
 
-When the popup appears, click `Install`.
+Upon running the command, you should see a a popup appear.
````
a a

jarolrod replied:
fixed

fanquake commented on doc/build-osx.md:
````diff
@@ -1,116 +1,319 @@
-## Dependencies
-```shell
-brew install automake libtool boost miniupnpc libnatpmp pkg-config python qt libevent qrencode
+Homebrew is a package manager for macOS that allows one to install packages from the command line easily.
````
I preferred the one liner with a link to https://docs.brew.sh/Installation. I don't think we need a paragraph explaining brew, or listing the actual install invocation.

jarolrod replied:
removed the install invocation in c180c911b88b2bd2baf2c9c2b24e276787ffb69b

I want this paragraph to just be a quick explanation that the commands use homebrew but they are free to use whatever package manager they use and adapt the commands. I added the following line to make this clearer:

```
Otherwise, you can adapt the commands to your package manager of choice.
```

fanquake reviewed (commented):
It seems sub optimal that we are basically just copy pasting blocks of dependencies throughout all the build docs. Isn't the point of dependencies.md to list the dependencies, what they are, and when they are required?

I think the same can be said for all of the "configuration" blocks. These seem to get repeated (across lots of docs) in varying combinations and levels of verbosity, and I'd think a lot of the time they might just be downright confusing to a first time builder.

This also includes the parts where configure output is basically pasted verbatim into documentation. i.e upnp or libnatpmp options. Can't we just tell users to look at `configure --help`? This means they'll always be getting accurate information; rather than something pasted into a build doc that now may have drifted from what actually happens.

jarolrod commented:
> It seems sub optimal that we are basically just copy pasting blocks of dependencies throughout all the build docs. Isn't the point of dependencies.md to list the dependencies, what they are, and when they are required?

Not exactly copy-pasting. Naming and links are different. I don't think it would be 'better' to point to dependencies.md and have the user figure out how that dependency correlates to their system.

> I think the same can be said for all of the "configuration" blocks. These seem to get repeated (across lots of docs) in varying combinations and levels of verbosity, and I'd think a lot of the time they might just be downright confusing to a first time builder.

Just trying to improve that situation across the build docs so that it's informative in the right ways to get someone up and running with a couple of examples that are relevant.

> This also includes the parts where configure output is basically pasted verbatim into documentation. i.e upnp or libnatpmp options. Can't we just tell users to look at `configure --help`? This means they'll always be getting accurate information; rather than something pasted into a build doc that now may have drifted from what actually happens.

Well I was thinking that the build docs can be a one-stop place on how to get up and running. I think it would be way more confusing for a first time builder to have to hope 3 different files and try to piece together the right information. 

The information doesn't need to drift away from what happening if it's maintained.

jarolrod commented:
Updated from 4780ea7 -> c180c91, rebased to resolve conflict with #21346 and addressed some of @fanquake suggestions.

changes:
- updated references of `qt` to `qt@5` 
- Removed confusing `Xcode Tools` dependency
- Removed specific configuration to port mapping and zmq in favor of pointing the user to check out `./configure -help`

**Justification:**
I guess, following from fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/21343#pullrequestreview-602468501) , this may need some more justification. I simply think that this adds value and the format can help a first-time builder get up and running. I certainly don't want to waste reviewer time or contribute something that would be an annoyance to maintain. If this PR falls into that, then it's ok if its closed.

fanquake reviewed (approved):
ACK c180c911b88b2bd2baf2c9c2b24e276787ffb69b - I still think these are getting too verbose and we're duplicating information all over the place; dependencies, configure options, combinations of options etc. However if people are happy to maintain them, I guess it's fine for now, and this revamping has already happened for some of the other build READMEs.
